bindings/pyroot/ROOT.py

```diff
@@ -383,6 +385,85 @@ def _TTreeAsMatrix(self, columns=None, exclude=None, dtype="double", return_labe
 if (sys.maxsize > 2**32): # https://docs.python.org/3/library/platform.html#cross-platform
     _root.CreateScopeProxy( "TTree" ).AsMatrix = _TTreeAsMatrix
 
+# RDataFrame.AsNumpy feature
+def _RDataFrameAsNumpy(df, columns=None, exclude=None):
+    """Read-out the RDataFrame as a collection of numpy arrays.
+
+    The values of the dataframe are read out as numpy array of the respective type
+    if the type is a fundamental type such as float or int. If the type of the column
+    is a complex type, such as your custom class or a std::array, the returned numpy
+    array contains Python objects of this type interpreted via PyROOT.
+
+    Be aware that reading out custom types is much less performant than reading out
+    fundamental types such as supported by directly by numpy.
+
+    Note that the reading is performed in multiple threads if the implicit
+    multi-threading of ROOT is enabled.
+
+    Parameters:
+        columns: If None return all branches as columns, otherwise specify names in iterable.
+        exclude: Exclude branches from selection.
+
+    Returns:
+        dict: Dict with column names as keys and 1D numpy arrays with content as values
+    """
+    # Import numpy lazily
+    try:
+        from numpy import array, empty, object
+    except:
+        raise ImportError("Failed to import numpy during call of RDataFrame.AsNumpy.")
+
+    # Import deepcopy
+    try:
+        from copy import deepcopy
+    except:
+        raise ImportError("Failed to import copy during call of RDataFrame.AsNumpy.")
+
+    # Find all column names in the dataframe if no column are specified
+    if not columns:
+        columns = [c for c in df.GetColumnNames()]
+
+    # Exclude the specified columns
+    if exclude == None:
+        exclude = []
+    columns = [col for col in columns if not col in exclude]
+
+    # Cast input node to base RNode type
+    df_rnode = _root.ROOT.Detail.RDF.AsRNode(df)
+
+    # Register Take action for each column
+    takes = {}
+    for column in columns:
+        column_type = df_rnode.GetColumnType(column)
+        takes[column] = _root.ROOT.Detail.RDF.RDataFrameTake(column_type)(df_rnode, column)
+
+    # Get values as C++ types
+    # These are not actual values but references, which need to be copied. Otherwise, these
+    # go out of scope if the Take node is deleted.
+    cpp_values = {}
+    for column in columns:
+        cpp_values[column] = takes[column].GetValue()
+
+    # Convert to numpy arrays
+    py_values = {}
+    for column in columns:
+        # The adoption mechanism through the __array_interface__ allows efficient copying without
+        # iteration over vectors of fundamental types. The copy is needed so that the memory
+        # is copied over and not adopted since the values returned by GetValue are only references.
+        if hasattr(cpp_values[column], "__array_interface__"):
+            py_values[column] = array(cpp_values[column], copy=True)
```

Comment:
It's not wasteful because the array interface allows numpy to make a bulk copy without any loops involved. But you got the point, we would need to return the `RResultPtr` so that the actual values on the C++ do not go out of scope.

The very good idea here is to make this a private member of the dictionary. I wanted to prevent that it appears in the actual returned value. But ofc I could attach it to it so that all of it goes out of scope at the same time.

However, this let's me wonder about such cases:

```
def foo:
    npy = ROOT.RDataFrame(4).Define("x", "42").AsNumpy()
    x = npy["x"]
    return x

print(x) # are the values of x still around or cleaned up?
```

Reply:
i suppose that `x` would be invalid out of the scope and you break with a segfault.

Reply:
I do not understand why. The RResultPtr in this case behaves like a wrapper around a shared_ptr which preserves ownership, but I may be missing something.

Reply:
The RResultPtrs go out of scope at the end of the `AsNumpy` function call, bringing the data with them.

Attaching them to the returned dictionary might break in cases such as the one @stwunsch points out, but we discussed this afternoon that attaching the RResultPtrs to their corresponding numpy arrays might actually work (I for one don't see a case in which it wouldn't, but it certainly requires careful consideration :smile: )

Reply:
as it seems it is not possible to inject new members to numpy arrays directly. we could write a wrapper python class around a numpy array to make this possible, but i assume that this is super confusing for the user:

```python
class NdarrayWrapper(numpy.ndarray):
    cpp_reference = None
    def__init__(self):
        super(NdarrayWrapper, self).__init__()
```

@etejedor Any better idea?

Reply:
There's even a wiki about subclassing ndarrays, seems pretty complicated solution: https://docs.scipy.org/doc/numpy-1.13.0/user/basics.subclassing.html

Reply:
from that page, adding an extra attribute seems well supported: https://docs.scipy.org/doc/numpy-1.13.0/user/basics.subclassing.html#slightly-more-realistic-example-attribute-added-to-existing-array

Reply:
But the user would see that it's a `NdarrayWrapper` type :thinking: Isn't this pretty confusing?

Reply:
But you are right, copying the base class they provided there, it does not seem that impossible.

Reply:
Ok, well. This works as object in the dictionary. It behaves like a numpy array even though the type tells you `NdarrayWrapper`. Actually, probably we can implement laziness here :thinking: 

```python
class NdarrayWrapper(ndarray):
        def __new__(cls, numpy_array, cpp_reference):
            obj = asarray(numpy_array).view(cls)
            obj.cpp_reference = cpp_reference
            return obj

        def __array_finalize__(self, obj):
            if obj is None: return
            self.cpp_reference = getattr(obj, 'cpp_reference', None)
```

Reply:
> But the user would see that it's a NdarrayWrapper type thinking Isn't this pretty confusing?

it's python, right? if it walks like a numpy array and quacks like a numpy array...

Reply:
It walks and quacks like a numpy array but probably likes to eat users alive :D

Reply:
We could call the class `ThisIsActuallyANumpyArray` :sweat_smile:

Reply:
Proposal will come soon!